internal/impl/mysql/snapshot.go

```diff
@@ -0,0 +1,216 @@
+// the License. You may obtain a copy of the License at
+//
+// https://github.com/redpanda-data/connect/v4/blob/main/licenses/rcl.md
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+// Snapshot represents a structure that prepares a transaction
+// and creates mysql consistent snapshot inside the transaction
+type Snapshot struct {
+	db *sql.DB
+	tx *sql.Tx
+
+	lockConn     *sql.Conn
+	snapshotConn *sql.Conn
+
+	logger *service.Logger
+	ctx    context.Context
+}
+
+// NewSnapshot creates new snapshot instance
+func NewSnapshot(ctx context.Context, logger *service.Logger, db *sql.DB) *Snapshot {
+	return &Snapshot{
+		db:     db,
+		ctx:    ctx,
+		logger: logger,
+	}
+}
+
+func (s *Snapshot) prepareSnapshot(ctx context.Context) (*mysql.Position, error) {
+	var err error
+	// Create a separate connection for FTWRL
+	s.lockConn, err = s.db.Conn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lock connection: %v", err)
+	}
+
+	// Create another connection for the snapshot
+	s.snapshotConn, err = s.db.Conn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create snapshot connection: %v", err)
+	}
+
+	// 1. Start a consistent snapshot transaction
+	s.tx, err = s.snapshotConn.BeginTx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to start transaction: %v", err)
+	}
+
+	// Execute START TRANSACTION WITH CONSISTENT SNAPSHOT
+	if _, err := s.tx.ExecContext(ctx, "START TRANSACTION WITH CONSISTENT SNAPSHOT"); err != nil {
+		if rErr := s.tx.Rollback(); rErr != nil {
+			return nil, rErr
+		}
+
+		return nil, fmt.Errorf("failed to start consistent snapshot: %v", err)
+	}
+
+	// 2. Acquire global read lock (minimizing lock time)
+	if _, err := s.lockConn.ExecContext(ctx, "FLUSH TABLES WITH READ LOCK"); err != nil {
+		if rErr := s.tx.Rollback(); rErr != nil {
+			return nil, rErr
+		}
+		return nil, fmt.Errorf("failed to acquire global read lock: %v", err)
+	}
+
+	// 3. Get binary log position (while locked)
+	pos, err := s.getCurrentBinlogPosition()
+	if err != nil {
+		// Make sure to release the lock if we fail
+		if _, eErr := s.lockConn.ExecContext(ctx, "UNLOCK TABLES"); eErr != nil {
+			return nil, eErr
+		}
+
+		if rErr := s.tx.Rollback(); rErr != nil {
+			return nil, rErr
+		}
+		return nil, fmt.Errorf("failed to get binlog position: %v", err)
+	}
+
+	// 4. Release the global read lock immediately
+	if _, err := s.lockConn.ExecContext(ctx, "UNLOCK TABLES"); err != nil {
+		if rErr := s.tx.Rollback(); rErr != nil {
+			return nil, rErr
+		}
+		return nil, fmt.Errorf("failed to release global read lock: %v", err)
+	}
+
+	return &pos, nil
+}
+
+func (s *Snapshot) getRowsCount(table string) (int, error) {
+	var count int
+	if err := s.tx.QueryRowContext(s.ctx, "SELECT COUNT(*) FROM "+table).Scan(&count); err != nil {
```

Comment:
Is this actually fast? Should we be querying the table stats instead? https://stackoverflow.com/a/61548683

Reply:
Why do we need a count anyways? In postgres we just query until we get no rows back, meaning we've reached the end of a table (because the pk is sorted). Can we do the same here? I think this will be slow and expensive for large tables.